test/e2e/network/loadbalancer.go

```diff
@@ -1192,6 +1198,9 @@ var _ = SIGDescribe("ESIPP [Slow]", func() {
 			}
 		}
 
+		// Look for EnsuredLoadBalancer event to signal that service changes have been processed by the service controller
+		// EnsuredLoadBalancer event will occur regardless of cloudprovider
+		WatchEvent(f, v1.EventTypeNormal, "EnsuredLoadBalancer", namespace, eventsList.ResourceVersion, loadBalancerCreateTimeout)
```

Comment:
Hmm - I'm actually not a fan of event-based approach. Events are best effort and don't have 100% guarantee to be delivered.
The other question is - IIRC the fact that the underlying ensureLoadBalancer() function is finished doesn't mean that something is necessary already programmed correctly. Happy to chat internally more about it.


The problem at large scale is that programming LB may actually take more then 5 minutes (KubeProxyLagTimeout below). In other places it seems that just increasing timeouts actually solved the problems.

So how about simply changing KubeProxyLagTimeout to GetServicePropagationTimeout:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/service/resource.go#L113

?

Reply:
In general, I do agree that an event based approach is not best. However in this situation the event seems like the closest signal we can get that changes have propagated.

My concern with extending the timeout is that though the test will become less flaky, it doesn't necessarily mean the changes were propagated. For example, if the load balancer processing is extremely slow, the entire test can complete even before the first change propagates to the load balancer. In the original [investigation](https://github.com/kubernetes/kubernetes/issues/93932#issuecomment-705889377) in #93932, some of the passed test cases were because the first deletion of the forwarding rule only occurred after the entire test passed. I think extending the timeout would lead to more false positives.

Reply:
One other approach could just to wait an amount of time after each service change which would make this test longer because of the guaranteed waits.

Reply:
If the test was passing even though the LB wasn't re-programmed correctly before, that clearly suggest that the test isn't well written.

We should come up with a way of testing that that succeeds when the traffic-policy is set to Cluster and doesn't succeed when it is set to Local and vice versa.

Your comment above shows that it's not really the case. And in that case, we shouldn't try to workaround that by relying on events, which first doesn't mean that it was fully programmed, second aren't guaranteed to be delivered.

What we should do is after switching the service to `Cluster` traffic-policy make a check that really works only if the policy is the `Cluster` policy.

@aojea @thockin - for thoughts

Reply:
> We should come up with a way of testing that that succeeds when the traffic-policy is set to Cluster and doesn't succeed when it is set to Local and vice versa.

Wow, so we don't have tests for these?
It is easy to implement, since with cluster you can poll the NodePort in ANY node of the cluster and with Local only in a node with Pods that belong to the service

You can poll it from inside one node of the cluster to make the test more portable, this is a quick example with KIND
```
$ kubectl get services
NAME               TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)        AGE
kubernetes         ClusterIP   10.96.0.1      <none>        443/TCP        70s
nodeport-service   NodePort    10.96.17.150   <none>        80:30274/TCP   9s

 
 $ kubectl get pods -o wide
NAME                               READY   STATUS    RESTARTS   AGE   IP           NODE           NOMINATED NODE   READINESS GATES
httpd-deployment-cc5f6b958-f5gbb   1/1     Running   0          43s   10.244.2.2   kind-worker2   <none>           <none>
httpd-deployment-cc5f6b958-kqq72   1/1     Running   0          43s   10.244.2.3   kind-worker2   <none>           <none>

kubectl get nodes -o wide
NAME                 STATUS   ROLES                  AGE     VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE       KERNEL-VERSION          CONTAINER-RUNTIME
kind-control-plane   Ready    control-plane,master   4m5s    v1.20.2   172.19.0.4    <none>        Ubuntu 20.10   4.18.0-259.el8.x86_64   containerd://1.4.0-106-gce4439a8
kind-worker          Ready    <none>                 3m31s   v1.20.2   172.19.0.3    <none>        Ubuntu 20.10   4.18.0-259.el8.x86_64   containerd://1.4.0-106-gce4439a8
kind-worker2         Ready    <none>                 3m31s   v1.20.2   172.19.0.2    <none>        Ubuntu 20.10   4.18.0-259.el8.x86_64   containerd://1.4.0-106-gce4439a8

 Type Cluster I reach it in any node
docker exec -it kind-worker2 curl --connect-timeout 2 172.19.0.4:30274
<html><body><h1>It works!</h1></body></html>
$ docker exec -it kind-worker2 curl --connect-timeout 2 172.19.0.2:30274
<html><body><h1>It works!</h1></body></html>
$ docker exec -it kind-worker curl --connect-timeout 2 172.19.0.2:30274
<html><body><h1>It works!</h1></body></html>
$ docker exec -it kind-worker curl --connect-timeout 2 172.19.0.4:30274
<html><body><h1>It works!</h1></body></html>


Change to Local
docker exec -it kind-worker2 curl --connect-timeout 2 172.19.0.4:30274
curl: (28) Connection timed out after 2001 milliseconds
docker exec -it kind-worker2 curl --connect-timeout 2 172.19.0.2:30274
<html><body><h1>It works!</h1></body></html>
```

Reply:
> Wow, so we don't have tests for these?

Technically this test is supposed to do that. But based on the comments I understand that it's not doing exactly what I would expect.

Reply:
I would add a test for service externaltraffic policy so the logic can be imported here,

Reply:
sorry, I got it wrong, there at least 2 test that check the externaltrafficpolicy feature https://github.com/kubernetes/kubernetes/blob/b0c4f5c6faff15e151440ca4d298d21979ccbfd9/test/e2e/network/loadbalancer.go#L958

https://github.com/kubernetes/kubernetes/blob/b0c4f5c6faff15e151440ca4d298d21979ccbfd9/test/e2e/network/loadbalancer.go#L926

Reply:
Sorry, let me clarify. When `ExternalTrafficPolicy=Cluster`, in this test, we do test internally on nodes that do not have a pod and verify that the request is successful: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/network/loadbalancer.go#L1161-L1166

The issue is when we test from the outside through the loadbalancer. From the outside there is no way to know how the loadbalancer has been programmed without having cloud provider specific code. The other non events approach I can think of is having mandatory waits after each change. However, that is no guarantee that the loadbalancer has been reprogrammed.